src/main/java/org/apache/commons/crypto/cipher/OpenSslGaloisCounterMode.java

```diff
@@ -0,0 +1,302 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.crypto.cipher;
+
+import javax.crypto.AEADBadTagException;
+import javax.crypto.BadPaddingException;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.GCMParameterSpec;
+import java.io.ByteArrayOutputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.spec.AlgorithmParameterSpec;
+
+/**
+ * This class do the real work(Encryption/Decryption/Authentication) for the authenticated mode: GCM.
+ *
+ * It calls the OpenSSL API to implement the JCE-like behavior
+ *
+ * @since 1.1
+ */
+class OpenSslGaloisCounterMode extends OpenSslFeedbackCipher {
+
+    // buffer for AAD data; if consumed, set as null
+    private ByteArrayOutputStream aadBuffer = new ByteArrayOutputStream();
```

Comment:
this should be set inside init() so the cipher with aad is reusable after clean(). [tested]